### ISSUE TITLE
parfold?

### ISSUE THREAD

**Issue by JuliaLawall** [UnixJunkie/parany#47]
Could Parmap's parfold be supported?  From the types, it looks like Parany.Parmap.parfold is Parmap's parmapfold.

**Comment by UnixJunkie** [comment]
Yes, Parany.Parmap.parfold is a parallel fold.
It doesn't have the exact same interface as Parmap.parmapfold, so I named it differently.
The interface I provide in parany is closer to List.fold.

**Comment by JuliaLawall** [comment]
Giving it the same name as another Parmap function that has an even more different interface is not a great solution.
It would be nice to have a Parany.Parmap.parfold that has the interface of the parmap one, so parmap users can just convert directly.

**Comment by UnixJunkie** [comment]
The Parany.Parmap module is a wrapper for "near compatibility" with parmap.
I don't intend to have all the parmap functions, with the exact same interface.
In the end, users of parany are supposed to use Parany.run: the most generic interface, under which it should be possible to code any parmap function.

In your project, you could have a module to easily switch between parmap and parany; but I am not sure I would want
to maintain such a module inside of parany; even if it came as an extenal contribution (more code -> more bugs -> more tests -> and more maintenance cost for me).

**Comment by JuliaLawall** [comment]
OK, even if you don't want to support the real Parmap.parfold. could you at least change the name?  I find parfold really hard to understand in the first place, and it is even more confusing to have something with the same name and a different type.

**Comment by UnixJunkie** [comment]
I am sorry, but I don't think I will change the name:
parallel map -> parmap
parallel iter -> pariter
parallel fold -> parfold
Those are very cool and cute names.
And, if we release a parany version where this name is changed, it might break some existing software...

**Comment by JuliaLawall** [comment]
On Fri, 19 Feb 2021, Francois Berenger wrote:

>
> I am sorry, but I don't think I will change the name:
> parallel map -> parmap
> parallel iter -> pariter
> parallel fold -> parfold
> Those are very cool and cute names.

I guess, but it seems very misleading to put it in the Parmap module.

And how is it a parallel fold?  To me a fold always has a function of type
a -> b -> b.  That is, it allows you to discard uninteresting values of a
immediately.  I don't understand how to use the parfold that is offered
(nor how to use Parmap.parmapfold).

julia

**Comment by UnixJunkie** [comment]
Ok, Parany.Parmap.parfold is honestly a parallel map.
The fold is not done in parallel by the workers (it is only done by the 'muxer' process I call it).

If you give me the type specification of the fold function that you want, I might send you an implementation, probably sometimes next week.

**Comment by JuliaLawall** [comment]
On Fri, 19 Feb 2021, Francois Berenger wrote:

>
> Ok, Parany.Parmap.parfold is honestly a parallel map.
> The fold is not done in parallel by the workers (it is only done by the 'muxer' process I call it).
>
> If you give me the type specification of the fold function that you want, I might send you an implementation, probably sometimes next week.

val parfold: ?init:(int -> unit) -> ?finalize:(unit -> unit) ->
?ncores:int -> ?chunksize:int -> ('a -> 'b -> 'b) -> 'a sequence -> 'b ->
('b->'b->'b) -> 'b
  (** [parfold ~ncores:n op (L l) b concat] computes [List.fold_right op l b]
      by forking [n] processes on a multicore machine.
      You need to provide the extra [concat] operator to combine the partial results of the
      fold computed on each core. If 'b = 'c, then [concat] may be simply [op].
      The order of computation in parallel changes w.r.t. sequential execution, so this
      function is only correct if [op] and [concat] are associative and commutative.
      If the optional [chunksize] parameter is specified,
      the processes compute the result in an on-demand fashion
      on blocks of size [chunksize].
      [parfold ~ncores:n op (A a) b concat] similarly computes [Array.fold_right op a b].
      *)

This is the documentation of parmap's parfold.  The thing that is
important to me is the function ('a -> 'b -> 'b), so I can get rid of
undesirable elements immediately.  I guess that with the current parfold,
one has to reprocess them in the combiner to drop the ones one doesn't
want.  That seems quite awkward.

julia

**Comment by UnixJunkie** [comment]
You can try using this in your project:
```
  let parfold_compat
      ?(init = fun (_rank: int) -> ()) ?(finalize = fun () -> ())
      ?(ncores: int option) ?(chunksize: int option) (f: 'a -> 'b -> 'b)
      (l: 'a list) (init_acc: 'b) (acc_fun: 'b -> 'b -> 'b): 'b =
    let nprocs = match ncores with
      | None -> 1 (* if the user doesn't know the number of cores to use,
                     we don't know better *)
      | Some x -> x in
    let csize = match chunksize with
      | None -> 1
      | Some x -> x in
    Parany.Parmap.parfold ~init ~finalize ~preserve:false ~core_pin:false ~csize nprocs
      (fun x -> f x init_acc) acc_fun init_acc l
```
I am not 100% sure this is the runtime behavior that you want though.
Also, you might want to change the default nprocs in this code, the ~preserve:false
and the ~core_pin:false. They are hard-coded defaults and some people might not like them.

**Comment by UnixJunkie** [comment]
If the code where you use Parmap.parfold is open-source, please let me know and I will go have a look at it.

**Comment by JuliaLawall** [comment]
On Fri, 19 Feb 2021, Francois Berenger wrote:

>
> If the code where you use Parmap.parfold is open-source, please let me know and I will go have a look at it.

https://github.com/hvdthong/PatchNetTool, preprocessing subdirectory.

The uses are in getinfo.ml and patch.ml.  Here is the code from
getinfo.ml:

      Parmap.parfold ~ncores:(!C.cores) ~chunksize:C.chunksize
        ~init:(fun id -> CD.me := id; Parmap.redirect ~path:prefix ~id)
        ~finalize:(fun () -> flush_all ())
        (fun ((commit,_,_,_,_,files) as x) rest ->
          Printf.eprintf "starting %s\n" commit; flush stderr;
          try
            match CD.getone (commit,files) with
              ([],_) -> rest
            | cur -> (x,cur) :: rest
          with CD.Failed -> rest)
        (Parmap.L infos) [] (@) in

thanks,
julia

**Comment by JuliaLawall** [comment]
On Fri, 19 Feb 2021, Francois Berenger wrote:

>
> You can try using this in your project:
>
>   let parfold_compat
>       ?(init = fun (_rank: int) -> ()) ?(finalize = fun () -> ())
>       ?(ncores: int option) ?(chunksize: int option) (f: 'a -> 'b -> 'b)
>       (l: 'a list) (init_acc: 'b) (acc_fun: 'b -> 'b -> 'b): 'b =
>     let nprocs = match ncores with
>       | None -> 1 (* if the user doesn't know the number of cores to use,
>                      we don't know better *)
>       | Some x -> x in
>     let csize = match chunksize with
>       | None -> 1
>       | Some x -> x in
>     Parany.Parmap.parfold ~init ~finalize ~preserve:false ~core_pin:false ~csize nprocs
>       (fun x -> f x init_acc) acc_fun init_acc l
>
> I am not 100% sure this is the runtime behavior that you want though.
> Also, you might want to change the default nprocs in this code, the ~preserve:false
> and the ~core_pin:false. They are hard-coded defaults and some people might not like them.

Thanks for the suggestion!  It is true that with this parfold behavior,
one doesn't have access to the full rest of the results in the a -> b -> b
function.  But in my case, and in many (but of course not all) cases where
I use fold, I don't want to actually look at that value, just add
something to it or not.

julia

**Comment by UnixJunkie** [comment]
I will send you a version which should parallelize better.

**Comment by UnixJunkie** [comment]
Another version of parfold_compat:
```
  let parfold_compat
      ?(init = fun (_rank: int) -> ()) ?(finalize = fun () -> ())
      ?(ncores: int option) ?(chunksize: int option) (f: 'a -> 'b -> 'b)
      (l: 'a list) (init_acc: 'b) (acc_fun: 'b -> 'b -> 'b): 'b =
    let nprocs = match ncores with
      | None -> 1 (* if the user doesn't know the number of cores to use,
                     we don't know better *)
      | Some x -> x in
    let csize = match chunksize with
      | None -> 1
      | Some x -> x in
    if nprocs <= 1 then
      List.fold_left (fun acc x -> f x acc) init_acc l
    else
      let input = ref l in
      let demux () = match !input with
        | [] -> raise End_of_input
        | _ ->
          let this_chunk, rest = BatList.takedrop csize !input in
          input := rest;
          this_chunk in
      let work xs =
        List.fold_left (fun acc x -> f x acc) init_acc xs in
      let output = ref init_acc in
      let mux x =
        output := acc_fun !output x in
      (* parallel work *)
      run ~init ~finalize
        (* leave csize=1 bellow *)
        ~preserve:false ~core_pin:false ~csize:1 nprocs ~demux ~work ~mux;
      !output
```
It should parallelize better than the previous one I sent (less marshal/unmarshal from the workers to the muxer process
if you use a chunksize > 1).
You might want to tweak the code, you need batteries because of BatList.takedrop that I don't want to reimplement.

**Comment by JuliaLawall** [comment]
On Tue, 23 Feb 2021, Francois Berenger wrote:

>
> Another version of parfold_compat:
>
>   let parfold_compat
>       ?(init = fun (_rank: int) -> ()) ?(finalize = fun () -> ())
>       ?(ncores: int option) ?(chunksize: int option) (f: 'a -> 'b -> 'b)
>       (l: 'a list) (init_acc: 'b) (acc_fun: 'b -> 'b -> 'b): 'b =
>     let nprocs = match ncores with
>       | None -> 1 (* if the user doesn't know the number of cores to use,
>                      we don't know better *)
>       | Some x -> x in
>     let csize = match chunksize with
>       | None -> 1
>       | Some x -> x in
>     if nprocs <= 1 then
>       List.fold_left (fun acc x -> f x acc) init_acc l
>     else
>       let input = ref l in
>       let demux () = match !input with
>         | [] -> raise End_of_input
>         | _ ->
>           let this_chunk, rest = BatList.takedrop csize !input in
>           input := rest;
>           this_chunk in
>       let work xs =
>         List.fold_left (fun acc x -> f x acc) init_acc xs in
>       let output = ref init_acc in
>       let mux x =
>         output := acc_fun !output x in
>       (* parallel work *)
>       run ~init ~finalize
>         (* leave csize=1 bellow *)
>         ~preserve:false ~core_pin:false ~csize:1 nprocs ~demux ~work ~mux;
>       !output
>
> It should parallelize better than the previous one I sent (less
> marshal/unmarshal from the workers to the muxer process
> if you use a chunksize > 1).
> You might want to tweak the code, you need batteries because of
> BatList.takedrop that I don't want to reimplement.

Thanks.  I have been using chunksize 1.  In the current setting, I'm a bit
hesitant to add more dependencies, but it could be useful in the future.

julia

**Comment by JuliaLawall** [comment]
On Tue, 23 Feb 2021, Francois Berenger wrote:

>
> Another version of parfold_compat:
>
>   let parfold_compat
>       ?(init = fun (_rank: int) -> ()) ?(finalize = fun () -> ())
>       ?(ncores: int option) ?(chunksize: int option) (f: 'a -> 'b -> 'b)
>       (l: 'a list) (init_acc: 'b) (acc_fun: 'b -> 'b -> 'b): 'b =
>     let nprocs = match ncores with
>       | None -> 1 (* if the user doesn't know the number of cores to use,
>                      we don't know better *)
>       | Some x -> x in
>     let csize = match chunksize with
>       | None -> 1
>       | Some x -> x in
>     if nprocs <= 1 then
>       List.fold_left (fun acc x -> f x acc) init_acc l
>     else
>       let input = ref l in
>       let demux () = match !input with
>         | [] -> raise End_of_input
>         | _ ->
>           let this_chunk, rest = BatList.takedrop csize !input in
>           input := rest;
>           this_chunk in
>       let work xs =
>         List.fold_left (fun acc x -> f x acc) init_acc xs in
>       let output = ref init_acc in
>       let mux x =
>         output := acc_fun !output x in
>       (* parallel work *)
>       run ~init ~finalize
>         (* leave csize=1 bellow *)
>         ~preserve:false ~core_pin:false ~csize:1 nprocs ~demux ~work ~mux;
>       !output
>
> It should parallelize better than the previous one I sent (less
> marshal/unmarshal from the workers to the muxer process
> if you use a chunksize > 1).
> You might want to tweak the code, you need batteries because of
> BatList.takedrop that I don't want to reimplement.

I used this code, with a definition of takedrop.  With some other fixes,
all seems to be well now.

julia

**Comment by UnixJunkie** [comment]
In your Parmap.parfold use example, you use flush_all as the finalize function.
Is this really necessary? Doesn't the OS already do this before terminating a process?
I understand the need to call flush_all() before starting parfold, but not so much after that.

**Comment by JuliaLawall** [comment]
On Thu, 25 Feb 2021, Francois Berenger wrote:

>
> In your Parmap.parfold use example, you use flush_all as the finalize
> function.
> Is this really necessary? Doesn't the OS already do this before terminating
> a process?
> I understand the need to call flush_all() before starting parfold, but not
> so much after that.

I seem to recall data that was not flushed.

julia

**Comment by UnixJunkie** [comment]
Ok, this is a special use case; a kind of parallel barrier because you pass data via the filesystem.
If you pass data between processes, make sure a (freshly written) file has been closed before you pass
its filename around.
